gradio/route_utils.py

```diff
@@ -244,3 +245,21 @@ async def call_process_api(
         output["data"] = output["data"][0]
 
     return output
+
+
+def set_replica_url_in_config(config: dict, replica_url: str) -> dict:
+    """
+    If the Gradio app is running on Hugging Face Spaces and the machine has multiple replicas,
+    we pass in the direct URL to the replica so that we have the fully resolved path to any files
+    on that machine. This direct URL can be shared with other users and the path will still work.
+    """
+    parsed_url = httpx.URL(replica_url)
+    stripped_url = parsed_url.copy_with(query=None)
+    stripped_url = str(stripped_url)
+    if not stripped_url.endswith("/"):
+        stripped_url += "/"
+
+    for component in config["components"]:
```

Comment:
Nice approach @abidlabs ! Very clean.

Two comments:
1. Should we be creating a new copy of the config here? We're modifying the config in place and we won't add the `root_url` if it's already present. So I think we're always going to get the first root_url that gets added
2. We still need to fix the client right? It's still configured to download from `/file=` and we don't look in the config to see if a root_url is set

Reply:
> Should we be creating a new copy of the config here? We're modifying the config in place and we won't add the root_url if it's already present. So I think we're always going to get the first root_url that gets added

Good point, I'll update in place instead

> We still need to fix the client right? It's still configured to download from /file= and we don't look in the config to see if a root_url is set

Nope the client doesn't need to know about this change. Since the Client's IP address is presumably fixed, it'll always get the correct asset. 

You can confirm with 

```py
from gradio_client import Client

client = Client("abidlabs/test-client-replica")
client.predict("abc")
```

Reply:
Updated the method to fix the dict in place and installed on the Space here: https://huggingface.co/spaces/abidlabs/test-client-replica

Reply:
+ added unit tests

Reply:
Actually fixed now @freddyaboulton